### PR TITLE
Add resource for GCP Static Account

### PR DESCRIPTION
**Pull request by lawliet89**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Implemented in https://github.com/hashicorp/vault-plugin-secrets-gcp/pull/107, merged to Vault in https://github.com/hashicorp/vault/pull/12023 and documented in https://github.com/hashicorp/vault/pull/12027

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-vault/blob/master/CHANGELOG.md):

```release-note
Add `vault_gcp_secret_static_account` resource
```

Output from acceptance testing:

Test was done by installing a custom build of the plugin into a Vault Dev server.

```
$ VAULT_ADDR=http://127.0.0.1:8200 VAULT_TOKEN=12345 GOOGLE_PROJECT=<snip> GOOGLE_CREDENTIALS=<snip> TESTARGS="--run TestGCPSecretStaticAccount" make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v --run TestGCPSecretStaticAccount -timeout 120m
?       github.com/hashicorp/terraform-provider-vault   [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/coverage      [no test files]
?       github.com/hashicorp/terraform-provider-vault/cmd/generate      [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/codegen   (cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/generated [no test files]
WARNING: Package "github.com/golang/protobuf/protoc-gen-go/generator" is deprecated.
        A future release of golang/protobuf will delete this package,
        which has long been excluded from the compatibility promise.

testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/decode    (cached) [no tests to run]
WARNING: Package "github.com/golang/protobuf/protoc-gen-go/generator" is deprecated.
        A future release of golang/protobuf will delete this package,
        which has long been excluded from the compatibility promise.

testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/datasources/transform/encode    (cached) [no tests to run]
WARNING: Package "github.com/golang/protobuf/protoc-gen-go/generator" is deprecated.
        A future release of golang/protobuf will delete this package,
        which has long been excluded from the compatibility promise.

testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/alphabet    (cached) [no tests to run]
WARNING: Package "github.com/golang/protobuf/protoc-gen-go/generator" is deprecated.
        A future release of golang/protobuf will delete this package,
        which has long been excluded from the compatibility promise.

testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/role        (cached) [no tests to run]
WARNING: Package "github.com/golang/protobuf/protoc-gen-go/generator" is deprecated.
        A future release of golang/protobuf will delete this package,
        which has long been excluded from the compatibility promise.

testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/template    (cached) [no tests to run]
WARNING: Package "github.com/golang/protobuf/protoc-gen-go/generator" is deprecated.
        A future release of golang/protobuf will delete this package,
        which has long been excluded from the compatibility promise.

testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/generated/resources/transform/transformation      (cached) [no tests to run]
?       github.com/hashicorp/terraform-provider-vault/schema    [no test files]
testing: warning: no tests to run
PASS
ok      github.com/hashicorp/terraform-provider-vault/util      (cached) [no tests to run]
WARNING: Package "github.com/golang/protobuf/protoc-gen-go/generator" is deprecated.
        A future release of golang/protobuf will delete this package,
        which has long been excluded from the compatibility promise.

=== RUN   TestGCPSecretStaticAccount
--- PASS: TestGCPSecretStaticAccount (23.78s)
PASS
ok      github.com/hashicorp/terraform-provider-vault/vault     23.791s

...
```

### CODE REVIEW

File: vault/resource_identity_group.go [hunk 1]
```diff
@@ -140,7 +140,6 @@ func identityGroupUpdateFields(d *schema.ResourceData, data map[string]interface
 			data["member_entity_ids"] = d.Get("member_entity_ids").(*schema.Set).List()
 			data["member_group_ids"] = d.Get("member_group_ids").(*schema.Set).List()
 
-
```
Comment (lawliet89) [inline vault/resource_identity_group.go:143]: Caused by `make fmt`.

**Comment by lawliet89** [comment]
Hey, @austingebauer, not sure fi you're the right person for this PR. This is the final bit in the "GCP static account saga".

**Comment by austingebauer** [comment]
Thanks for doing this, @lawliet89! I will have a look at this along with others who are more familiar with the provider.

File: vault/resource_gcp_secret_static_account.go [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+		data["token_scopes"] = v.(*schema.Set).List()
+	}
+
+	if v, ok := d.GetOk("binding"); ok {
+		bindingsHCL := gcpSecretRenderBindingsFromData(v)
+		log.Printf("[DEBUG] Rendered GCP Secrets backend static account bindings HCL:\n%s", bindingsHCL)
+		data["bindings"] = bindingsHCL
+	} else {
+		data["bindings"] = ""
+	}
```
Comment (austingebauer) [inline vault/resource_gcp_secret_static_account.go:224]: Will the addition of this else block result in different behavior wrt `bindings` for static accounts and rolesets? I see that it's absent for rolesets ([resource_gcp_secret_roleset.go#L267](https://github.com/hashicorp/terraform-provider-vault/blob/master/vault/resource_gcp_secret_roleset.go#L267)).

File: vault/resource_gcp_secret_static_account.go [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+
+	log.Printf("[DEBUG] Writing GCP Secrets backend static account %q", path)
+
+	data := map[string]interface{}{}
+	gcpSecretStaticAccountUpdateFields(d, data)
+	d.SetId(path)
+	_, err := client.Logical().Write(path, data)
+	if err != nil {
+		d.SetId("")
+		return fmt.Errorf("Error writing GCP Secrets backend static account %q: %s", path, err)
```
Comment (austingebauer) [inline vault/resource_gcp_secret_static_account.go:114]:
```suggestion
		return fmt.Errorf("error writing GCP Secrets backend static account %q: %s", path, err)
```

File: vault/resource_gcp_secret_static_account.go [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+	path := d.Id()
+
+	data := map[string]interface{}{}
+	gcpSecretStaticAccountUpdateFields(d, data)
+
+	log.Printf("[DEBUG] Updating GCP Secrets backend static account %q", path)
+
+	_, err := client.Logical().Write(path, data)
+	if err != nil {
+		return fmt.Errorf("Error updating GCP Secrets backend static account %q: %s", path, err)
```
Comment (austingebauer) [inline vault/resource_gcp_secret_static_account.go:184]:
```suggestion
		return fmt.Errorf("error updating GCP Secrets backend static account %q: %s", path, err)
```

File: vault/resource_gcp_secret_static_account.go [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+}
+
+func gcpSecretStaticAccountDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	path := d.Id()
+
+	log.Printf("[DEBUG] Deleting GCP secrets backend static account %q", path)
+	_, err := client.Logical().Delete(path)
+	if err != nil {
+		return fmt.Errorf("Error deleting GCP secrets backend static account %q", path)
```
Comment (austingebauer) [inline vault/resource_gcp_secret_static_account.go:198]:
```suggestion
		return fmt.Errorf("error deleting GCP secrets backend static account %q", path)
```

File: vault/resource_gcp_secret_static_account.go [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+						},
+					},
+				},
+			},
+			"service_account_project": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Project of the GCP Service Account managed by this static account",
+			},
+		},
```
Comment (austingebauer) [inline vault/resource_gcp_secret_static_account.go:94]: Do you know if the [CustomizeDiff](https://github.com/hashicorp/terraform-provider-vault/blob/master/vault/resource_gcp_secret_roleset.go#L105-L114) used for rolesets is relevant static accounts?

File: website/docs/r/gcp_secret_static_account.html.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
+---
+layout: "vault"
+page_title: "Vault: vault_gcp_secret_static_account resource"
+sidebar_current: "docs-vault-resource-gcp-secret-statoc-account"
```
Comment (austingebauer) [inline website/docs/r/gcp_secret_static_account.html.md:4]:
```suggestion
sidebar_current: "docs-vault-resource-gcp-secret-static-account"
```

File: website/docs/r/gcp_secret_static_account.html.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+The following arguments are supported:
+
+* `backend` - (Required, Forces new resource) Path where the GCP Secrets Engine is mounted
+
+* `static_account` - (Required, Forces new resource) Name of the Static Account to create
+
+* `service_account_email` - (Required, Forces new resource) Email of the GCP service account to manage.
+
+* `secret_type` - (Optional, Forces new resource) Type of secret generated for this role set. Accepted values: `access_token`, `service_account_key`. Defaults to `access_token`.
```
Comment (austingebauer) [inline website/docs/r/gcp_secret_static_account.html.md:57]:
```suggestion
* `secret_type` - (Optional, Forces new resource) Type of secret generated for this static account. Accepted values: `access_token`, `service_account_key`. Defaults to `access_token`.
```

File: website/docs/r/gcp_secret_static_account.html.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+* `token_scopes` - (Optional, Required for `secret_type = "access_token"`) List of OAuth scopes to assign to `access_token` secrets generated under this role set (`access_token` role sets only).
```
Comment (austingebauer) [inline website/docs/r/gcp_secret_static_account.html.md:59]:
```suggestion
* `token_scopes` - (Optional, Required for `secret_type = "access_token"`) List of OAuth scopes to assign to `access_token` secrets generated under this static account (`access_token` static accounts only).
```

**Review by austingebauer — COMMENTED** [review]
Just a few comments and questions to get started. Thanks again, @lawliet89!

File: vault/provider.go [hunk 10]
```diff
@@ -427,6 +427,10 @@ var (
 			Resource:      gcpSecretRolesetResource(),
 			PathInventory: []string{"/gcp/roleset/{name}"},
 		},
+		"vault_gcp_secret_static_account": {
+			Resource:      gcpSecretStaticAccountResource(),
+			PathInventory: []string{"/gcp/static/{name}"},
```
[the remark this answers is not in the document]
Comment (lawliet89) [reply vault/provider.go:432]: Oops. Missed this.

File: vault/resource_gcp_secret_roleset_test.go [hunk 11]
```diff
@@ -57,10 +57,10 @@ func TestGCPSecretRoleset(t *testing.T) {
 				),
 			},
 			{
-				ResourceName:            "vault_gcp_secret_backend.test",
+				ResourceName:            "vault_gcp_secret_roleset.test",
 				ImportState:             true,
 				ImportStateVerify:       true,
-				ImportStateVerifyIgnore: []string{"credentials"},
+				ImportStateVerifyIgnore: []string{},
```
[the remark this answers is not in the document]
Comment (lawliet89) [reply vault/resource_gcp_secret_roleset_test.go:63]: I wrote the original test code for the roleset resource. This change is to correct the mistake. It should be testing the import for `vault_gcp_secret_roleset` and not `vault_gcp_secret_backend`. The roleset resource does not have a credentials field.

File: website/docs/r/gcp_secret_static_account.html.md [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+Each [static account](https://www.vaultproject.io/docs/secrets/gcp/index.html#static-accounts) is tied to a separately managed
+Service Account, and can have one or more [bindings](https://www.vaultproject.io/docs/secrets/gcp/index.html#bindings) associated with it.
+
+## Example Usage
+
+```hcl
+resource "google_service_account" "this" {
+  account_id = "my-awesome-account"
+}
```
[the remark this answers is not in the document]
Comment (lawliet89) [reply website/docs/r/gcp_secret_static_account.html.md:21]: Yes. It's quite common to use multiple providers, though. The `google` prefix in the resource name indicates it's from a different provider., It should be quite obvious. Do you think we should still add that comment?

File: vault/resource_gcp_secret_static_account.go [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+		return "", fmt.Errorf("no backend found")
+	}
+	res := gcpSecretStaticAccountBackendFromPathRegex.FindStringSubmatch(path)
+	if len(res) != 2 {
+		return "", fmt.Errorf("unexpected number of matches (%d) for backend", len(res))
+	}
+	return res[1], nil
+}
+
+func gcpSecretRoleSetdStaticAccountNameFromPath(path string) (string, error) {
```
[the remark this answers is not in the document]
Comment (lawliet89) [reply vault/resource_gcp_secret_static_account.go:254]: Sorry that was a renaming fail.

File: vault/resource_gcp_secret_static_account.go (hunk 2 above)
Comment (lawliet89) [reply vault/resource_gcp_secret_static_account.go:224]: `bindings` is required for rolesets but optional for static accounts. Thus the setting to `""`.

File: vault/resource_gcp_secret_roleset_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (austingebauer) [reply vault/resource_gcp_secret_roleset_test.go:63]: Makes sense, thanks!

File: website/docs/r/gcp_secret_static_account.html.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (austingebauer) [reply website/docs/r/gcp_secret_static_account.html.md:21]: Thanks for explaining! I think it's okay to leave it as it currently is then 👍

File: vault/resource_gcp_secret_static_account.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (austingebauer) [reply vault/resource_gcp_secret_static_account.go]: Ty!

File: vault/resource_gcp_secret_static_account.go (hunk 2 above)
Comment (austingebauer) [reply vault/resource_gcp_secret_static_account.go:224]: Ah, right. Thanks!

File: vault/resource_gcp_secret_static_account.go (hunk 6 above)
Comment (austingebauer) [reply vault/resource_gcp_secret_static_account.go:94]: Got it. Thanks for the explanation!

**Comment by austingebauer** [comment]
@lawliet89 - This is looking good. I'm going to have someone more familiar with the provider take a look before I merge it. Thanks again!

File: vault/gcp.go [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+import (
+	"bytes"
+	"fmt"
+	"sort"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/hashcode"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+type GCPBinding struct {
```
Comment (benashz) [inline vault/gcp.go:12]: I think it would be good to add some documentation here.

File: vault/resource_gcp_secret_static_account.go [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+
+	log.Printf("[DEBUG] Read GCP Secrets backend static account %q", path)
+	if resp == nil {
+		log.Printf("[WARN] GCP Secrets backend static account %q not found, removing from state", path)
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("backend", backend)
+	d.Set("static_account", staticAccount)
```
Comment (benashz) [inline vault/resource_gcp_secret_static_account.go:149]:
```suggestion
	if err := d.Set("backend", backend); err != nil {
		return err
	}
	if err := d.Set("static_account", staticAccount); err != nil {
		return err
	}
```

File: vault/resource_gcp_secret_static_account_test.go [hunk 16]
```diff
@@ -0,0 +1,397 @@
...
+					resource.TestCheckResourceAttr("vault_gcp_secret_static_account.test", fmt.Sprintf("binding.%d.resource", keyHash), fmt.Sprintf("//cloudresourcemanager.googleapis.com/projects/%s", project)),
+					resource.TestCheckResourceAttr("vault_gcp_secret_static_account.test", fmt.Sprintf("binding.%d.roles.#", keyHash), "1"),
+					resource.TestCheckResourceAttr("vault_gcp_secret_static_account.test", fmt.Sprintf("binding.%d.roles.2133424675", keyHash), updatedRole),
+				),
+			},
+		},
+	})
+}
+
+func testGCPSecretStaticAccount_attrs(backend, staticAccount string) resource.TestCheckFunc {
```
Comment (benashz) [inline vault/resource_gcp_secret_static_account_test.go:126]: I wonder if there is way we could perhaps refactor this test function to other, more discrete functions. As-is, it very thorough, but rather challenging to consume in on go :)

File: vault/gcp.go (hunk 14 above)
Comment (lawliet89) [reply vault/gcp.go:15]: Added something simple.

File: vault/resource_gcp_secret_static_account_test.go (hunk 16 above)
Comment (lawliet89) [reply vault/resource_gcp_secret_static_account_test.go:126]: This "pattern" is used throughout the tests in the provider which I mimicked here.. Perhaps a cleanup/refactor PR further down the line?

Comment (benashz) [reply vault/resource_gcp_secret_static_account_test.go:126]: Okay, that sounds fine, since a precedent has already been set.

**Review by benashz — APPROVED** [review]
LGTM!

**Review by austingebauer — APPROVED** [review]
LGTM!

**Pull request merged**